packages/viewer/src/components/viewer-markup/viewer-markup.tsx

```diff
@@ -0,0 +1,382 @@
+
+@Component({
+  tag: 'vertex-viewer-markup',
+  styleUrl: 'viewer-markup.css',
+  shadow: true,
+})
+export class ViewerMarkup {
+  /**
+   * An HTML template that describes the HTML to use for new arrow
+   * markup. It's expected that the template contains a
+   * `<vertex-viewer-markup-arrow>`.
+   */
+  @Prop()
+  public arrowTemplateId?: string;
+
+  /**
+   * An HTML template that describes the HTML to use for new circle
+   * markup. It's expected that the template contains a
+   * `<vertex-viewer-markup-circle>`.
+   */
+  @Prop()
+  public circleTemplateId?: string;
+
+  /**
+   * The type of markup to perform.
+   */
+  @Prop()
+  public tool: ViewerMarkupToolType = 'arrow';
+
+  /**
+   * If `true`, disables adding or editing of markup through user
+   * interaction.
+   */
+  @Prop()
+  public disabled = false;
+
+  /**
+   * The viewer to connect to markup. If nested within a <vertex-viewer>,
+   * this property will be populated automatically.
+   */
+  @Prop()
+  public viewer?: HTMLVertexViewerElement;
+
+  /**
+   * The ID of the markup that is selected.
+   */
+  @Prop({ mutable: true })
+  public selectedMarkupId?: string;
+
+  /**
+   * Dispatched when a new markup is added, either through user interaction
+   * or programmatically.
+   */
+  @Event()
+  public markupAdded!: EventEmitter<
+    HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+  >;
+
+  /**
+   * Dispatched when a markup is removed, either through user
+   * interaction or programmatically.
+   */
+  @Event()
+  public markupRemoved!: EventEmitter<
+    HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+  >;
+
+  @Element()
+  private hostEl!: HTMLElement;
+
+  /**
+   * Adds a new markup as a child to this component. A new markup
+   * component will be created from the template specified by
+   * `arrow-template-id`, `circle-template-id`, or if undefined
+   * a default element will be created.
+   *
+   * @param markup The markup to add.
+   * @returns The markup element that was created.
+   * @see {@link ViewerMarkups.arrowTemplateId}
+   * @see {@link ViewerMarkups.circleTemplateId}
+   */
+  @Method()
+  public async addMarkup(
+    markup: Markup
+  ): Promise<
+    HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+  > {
+    if (markup instanceof ArrowMarkup) {
+      const { start, end, id } = markup;
+
+      const el = this.createArrowMarkupElement();
+      el.id = id;
+      el.start = start;
+      el.end = end;
+
+      this.updatePropsOnMarkup(el);
+      this.hostEl.appendChild(el);
+      this.markupAdded.emit(el);
+      return el;
+    } else if (markup instanceof CircleMarkup) {
+      const { bounds, id } = markup;
+
+      const el = this.createCircleMarkupElement();
+      el.id = id;
+      el.bounds = bounds;
+
+      this.updatePropsOnMarkup(el);
+      this.hostEl.appendChild(el);
```

Comment:
Missing a `this.markupAdded.emit(el);`. Might be a good opportunity to pull this shared code into a function.